### PR TITLE
Deduplicate Strings in REST Bulk Request Parsing

### PR DESCRIPTION
**Pull request by original-brownbear**
We can save a little memory here since these strings might live for quite a while on the coordinating node.

### CODE REVIEW

**Comment by rjernst** [comment]
How much memory are we potentially talking about? While this change is not complex, which strings were chosen seems a bit arbitrary, and doing this only inside bulk requests seems like it leaves lots of places in other parts of the codebase that similar duplication would still exist.

**Comment by original-brownbear** [comment]
@rjernst 

> How much memory are we potentially talking about?

Depends on a lot of factors obviously. But thinking about use cases like a Logstash node using a 50k bulk size (say 100 bytes per index name -> 5MB of heap per request saved) I figured it was worth it.

>  only inside bulk requests seems like it leaves lots of places in other parts of the codebase that similar duplication would still exist.

Already dealt with part of this in the TCP communication in https://github.com/elastic/elasticsearch/pull/56094 and have a PR open for more parts in https://github.com/elastic/elasticsearch/pull/56209. 
Obviously the impact on the TCP layer is a lot more significant (because we save the `Index` instance serialization including the index uuids and so on) but I figured saving a few MB here "for ~free" can't hurt either?

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java [hunk 1]
```diff
@@ -115,6 +118,7 @@ public void parse(
         int line = 0;
         int from = 0;
         byte marker = xContent.streamSeparator();
+        final Map<String, String> stringDeduplicator = new HashMap<>();
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:121]: If we are going to do this, can we at least add a comment here explaining the intent? Specifically noting that this doesn't save on the actual instantiation of the duplicate strings, only on the lifetime of those duplicates (can be GC'd after parsing is complete).

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:124]: Sure thing, done in 02e3fc42f7b

**Review by rjernst — APPROVED** [review]
LGTM. Can we also add a test in BulkRequestParserTests, which can check reference equality of each of the values being deduplicated here?

**Comment by original-brownbear** [comment]
Thanks @rjernst ! Added a test in 7a6e9afbb6a

**Pull request merged**